src/Underworld.ts

```diff
@@ -2097,21 +2117,261 @@ export default class Underworld {
     );
     return { x, y };
   }
+  // Handles level completion, game over, turn phases, and hotseat
+  async progressGameState() {
+    console.trace('[GAME] Progress Game State...');
+
+    // Game State should not progress if no players are connected
+    const connectedPlayers = this.players.filter(p => p.clientConnected);
+    if (connectedPlayers.length == 0) {
+      console.log('[GAME] No connected players: ', this.players);
+      return;
+    }
+
+    // We should try progressing the level before ending the game
+    // in case the player has beaten the level and died at the same time
+    // Favoring the player in this scenario should only improve player experience
+    if (this.handleLevelProgress()) {
+      // Level progress has been handled by the above function
+      // So there is nothing left to do, just continue
+    }
+    else if (this.isGameOver()) {
+      this.doGameOver();
+    }
+    else {
+      console.log('[GAME] Turn Phase\nCurrent == ', turn_phase[this.turn_phase]);
+
+      // TODO - It would be cleaner if we found a way to move all TurnPhase logic here
+
+      // Most turn phases are currently handled in InitializeTurnPhase()
+      // the Player Turn depends on player input and thus happens asyncronously
+      // So we make a special case to handle it here instead
+      if (this.turn_phase == turn_phase.PlayerTurns) {
+        // If all hotseat players are ready, try end player turn
+        await this.handleNextHotseatPlayer();
+
+        // Moves to NPC.ALLY Phase if possible
+        await this.tryEndPlayerTurnPhase();
+      }
+
+      console.log('[GAME] Turn Phase\nNew == ', turn_phase[this.turn_phase]);
+    }
+
+    console.log('[GAME] Progress Game State Complete');
+    return;
+  }
+  handleLevelProgress(): boolean {
+    // Returns true if this function progresses the level state
+    // - Spawn the next wave of enemies
+    // - Spawn purple portals
+    // - Send players to next level
+
+    const connectedPlayers = this.players.filter(p => p.clientConnected);
+
+    // TODO - Below is a temp failsafe. It should be removed:
```

Comment:
Can you describe the conditions under which it is safe to remove?

Reply:
I'll clarify the TODO a bit 👍 

The failsafe can't be removed yet due to some other weird behaviour going on, but in theory there's no reason it should be there, since it's just accounting for an edge case being generated elsewhere in code. I think it has to do with the progressGameState() logic running on underworld generation / in headless when all players are still in lobby, but I haven't had the time to really investigate it.

Reply:
Yeah, it's there to account for an edgecase where progressGameState() is being called before it should be. Updated comment:

> // TODO - Below is a temp failsafe. It should be removed eventually:
// This failsafe is here to account for an edge case where progressGameState()
// is called during underworld generation / in headless when all players are still in lobby
// I think. Haven't had time to investigate.
// Point: Progress game state should not be getting called before the game "starts"
// I.E. level is generated, enemies are spawned, players are ready to play and spawn, etc.